feat: change set support in chat input and chat model

planger opened the pull request:
#### What it does

* Add the concept of a change set to chat model and input UI
* Add an implementation of change set elements for files
* Add an agent for testing: @ChangeSet

Other fixes in Chat Input UI:
* The inProgress state of the chat input was actually unsafely managed. This change addresses the proper management of the inProgress state.
* The positioning, e.g. of the placeholder is now more adaptive.

As the change set feature directly relates to another feature (context, work in progress), this change also already prepares for those changes in the chat UI:
* Prepare chat input for adding context to requests
* Add context in the form of variables to chat model

Fixes https://github.com/eclipse-theia/theia/issues/14749

#### How to test

To demonstrate the usage and simplify testing, this PR includes a test agent named `@ChangeSet`. This statically sets a somewhat random change set into the chat model, with one new, one modified, and one deleted file change set element.

https://github.com/user-attachments/assets/fd4347ec-4ed4-4808-ac41-3e4b4b0735c2

#### Follow-ups

Next steps include:

- [x] Hook up the [Coder Agent](https://github.com/eclipse-theia/theia/blob/master/packages/ai-workspace-agent/src/browser/coder-agent.ts) to allow for testing this change set support in more breadth.
- [ ] Allow to clear a changeset (all changes a single changes)
- [ ] When using with the Coder agent, the progress could be shown in a better way. We should add the change earlier while the LLM is still streaming. We might also open the file and show some line animation there.
* The [Coder Agent](https://github.com/eclipse-theia/theia/blob/master/packages/ai-workspace-agent/src/browser/coder-agent.ts) should be switched to use the monaco model, if it exists, to capture potentially dirty states -- otherwise the AI sees a different version than the user.
* Improve the diff editor we use here: currently there seems to be no simple way to show a diff between the workspace (editable) and a suggested modified version, while allowing to "accept" changes from the suggested version to the workspace version. Therefore, we take a somewhat ugly shortcut in this PR to show a diff between the suggested changes as readonly original version left and the editable workspace version on the right. This works, however, the colors are wrong (red for additions) and the labels are wrong (reject, even though it is actually an accept).
* Potentially optimize the `ChangeSetFileElement` how it applies and reverts changes, taking the monaco text model into account. I aimed at reusing the functionality to modify the monaco model if an editor is open without saving, and saving to disk directly if no editor is opened. But this should be tested for user experience and consistency.

#### Breaking changes

- [ ] This PR introduces breaking changes and requires careful review. If yes, the breaking changes section in the [changelog](https://github.com/eclipse-theia/theia/blob/master/CHANGELOG.md) has been updated.

#### Attribution

None

#### Review checklist

- [X] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

JonasHelming commented on examples/api-samples/src/browser/chat/change-set-chat-agent-contribution.ts:
```diff
@@ -0,0 +1,132 @@
+        request.response.response.addContent(new MarkdownChatResponseContentImpl(
+            'I have created a change set for you. You can now review and apply it.'
+        ));
+        request.response.complete();
```
I wonder when we should update the change set in a real scenario. Simple solution would be when a response is there, but if more then one file is changed, it looks too late from a user POV. Unfortunately, the agent currently does not know which functions are called, therefore, we have no good trigger.

planger replied:
As discussed, the agent can at any time update the chat model, and after https://github.com/eclipse-theia/theia/pull/14751 also tool functions can update it.

JonasHelming commented:
Very nice, looks really great. I believe this is a very powerful feature!

Couple of high level comments/questions:

- What is our story to close a changeset or to clear it?

JonasHelming commented on packages/ai-chat/src/browser/change-set-file-element.ts:
```diff
@@ -0,0 +1,224 @@
+        }
+    }
+
+    async reject(): Promise<void> {
```
'reject' is rather 'undo' or 'discard', two comments:
- I would rename this action and only show it, if the change has already been applied
- I believe we need something to remove a change from a changeset without applying it.

planger replied:
I see, so I would propose to add a `remove` action which is always visible and show a `discard` if the change has already been applied. WDYT?

JonasHelming replied:
yes, that makes sense. I would also add a "clear" action next to accept to clear the complete changeset

JonasHelming commented on packages/ai-chat/src/common/chat-model.ts:
```diff
@@ -420,21 +512,66 @@ export class ChatModelImpl implements ChatModel {
+    }
+
+    addElement(element: ChangeSetElement): void {
+        this.addElements([element]);
```
Why is no notification triggered in this case?

eneufeld replied:
this just delegates to the addElements method below, which triggers the notification

JonasHelming commented on packages/ai-chat/src/browser/change-set-file-element.ts:
```diff
@@ -0,0 +1,224 @@
+};
+
+@injectable()
+export class ChangeSetFileElement implements ChangeSetElement {
```
I am not 100% sure if I like the mixture between behavior and state in this class. It relies heavily on DI and Theia services and seems hard to test. Maybe we should extract the behavior to a separate interface?

JonasHelming replied:
My suggestion would be more to extract the behavior into a separate interface (delegation), I am not 100% sure myself, though.

JonasHelming reviewed (commented):
Just a first highlevel look for now

JonasHelming commented on packages/ai-chat/src/browser/change-set-file-resource.ts:
Is this used somewhere?

planger replied:
Yes, the resource resolver is registered here:
https://github.com/eclipse-theia/theia/pull/14750/files/0817dcbf334ccfe0a430798ffcbce1dbcce9baa9#diff-444a8731310f8ef30095775299ab7f86f4c392bdd312ac603b9b795f9b3759d4R103

This is needed to obtain the suggested content of a file via the Theia URIs resolution, enabling to reuse the DiffUris in Theia.

sdirix commented on packages/ai-chat-ui/src/browser/chat-input-widget.tsx:
```diff
@@ -215,15 +222,40 @@ const ChatInput: React.FunctionComponent<ChatInputProperties> = (props: ChatInpu
+    React.useEffect(() => {
+        const onChatModelUpdate = (event: ChatChangeEvent) => {
+            if (ChatChangeEvent.isChangeSetEvent(event)) {
+                forceUpdate();
```
It seems the `forceUpdate` is used to update the UI for changes in the changeSet. A bit more elegant would be to maintain a "view model" abstraction in `useState` for the `ChangeSetBox` and update that one whenever needed. Then we would not need a `forceUpdate`

sdirix commented on packages/ai-chat-ui/src/browser/chat-input-widget.tsx:
```diff
@@ -215,15 +222,40 @@ const ChatInput: React.FunctionComponent<ChatInputProperties> = (props: ChatInpu
+        };
+        const listener = lastRequest?.response.onDidChange(onResponseUpdate);
         return () => listener?.dispose();
     }, [lastRequest]);
```
This was not introduced by your PR but as I am just noticing it: We should never build `useEffect` based transformations for `state` which is under our control. At best this introduces unnecessary rerenderings and confusing state updates, at worst this ends up in endless rerender cycles.

As `lastRequest` is maintained by us, we should perform any side effects we want to see at the same time as `lastRrequest` is changed, i.e. by the code calling the setter.

sdirix commented on packages/ai-chat-ui/src/browser/chat-input-widget.tsx:
```diff
@@ -268,30 +300,143 @@ const ChatInput: React.FunctionComponent<ChatInputProperties> = (props: ChatInpu
         }
     };
 
+    const leftOptions = React.useMemo<Option[]>(() => (
+        props.showContext ? [{
+            title: 'Attach elements to context',
+            handler: () => { /* TODO */ },
+            className: 'codicon-add'
+        }] : []
+    ), [props.showContext]);
+
+    const rightOptions = React.useMemo<Option[]>(() => (
```
At the moment the memos are not worth it. The only thing we save are the construction of the arrays, however the `ChatInputOptions` will be rerendered anyway.

Only if we would `memo` the whole `ChatInputOptions`, then the `useMemo`s are needed. Not sure whether all these complications are worth it, I would imagine that the `ChatInputOptions` are cheap to render.

JonasHelming commented:
It is now integrated with Coder, two example prompts to get all three changes types:

@Coder Move the readme file to a sub direcotry called /readme

@Coder Implement the interface MCPServerManager in the same file (packages/ai-mcp/src/common/mcp-server-manager.ts). Don't ask me anything, jut do it.

planger commented:
@JonasHelming @sdirix 

I've addressed the review feedback and added delete functionality.

@sdirix I've reintroduce the `latestRequest` ref to have proper value in functions. Maybe there is a better solution `latestRequest` was always `undefined` when used in the function `onDeleteChangeSet` and `onDeleteChangeSetElement`.

JonasHelming commented:
Looks great!

I found one additional issue, either to fix now or as a follow up:
- Discard does not work for me after applying a change (accept a change => discard => nothing happens)

Once we merge this, we should create a follow up epic with all the open topics => I can do this.
